src/_pytest/fixtures.py

```diff
@@ -1056,9 +1056,17 @@ def execute(self, request: SubRequest) -> FixtureValue:
         my_cache_key = self.cache_key(request)
         if self.cached_result is not None:
             cache_key = self.cached_result[1]
-            # note: comparison with `==` can fail (or be expensive) for e.g.
-            # numpy arrays (#6497).
-            if my_cache_key is cache_key:
+
+            # First attempt to use 'is' for performance reasons (for example numpy arrays (#6497)).
```

Comment:
> I think this actually muddies the semantic water somewhat. It is possible for an object to not equal itself (i.e. obj is obj is true, but obj == obj is false).

While I agree this is possible, we can then argue with people that maybe they shouldn't be using such odd objects as parametrization values.

TBH I think the current implementation strikes a good balance between doing what is right with also doing some optimization... it is hard to argue that if we pass the same object (`is`) it *should not* use the same cache value.

Reply:
> While I agree this is possible, we can then argue with people that maybe they shouldn't be using such odd objects as parametrization values.

Actually, it does not matter if the object is equal or not to itself, but if given this object and their cache key, for the purposes of comparing their cache key and reuse the fixture value, are they the same?

So we do not need to concern ourselves with the correctness of comparison here, only how it affects the cache verification. For example, `nan` is famously an object with weird semantics (it is not equal to itself). But we do not need to argue if this is right or wrong, just to answer the question: if we have used `nan` for a cache key before, and now we also are comparing to `nan`, should we use the cached value or not? To me the answer is yes, even thought `nan != nan`.

So again I think this implementation now is correct.

Reply:
@nicoddemus Good comment, pathological cases like `nan` and its inverse `unitest.mock.ANY` (compares equal to anything) is what I had in mind.

Let me try to list the possible semantics I can think of:

1. `a is b` -- current semantics
2. `a == b` without fallback -- params must be comparable/hashable (error)
3. `a == b` with fallback to false
4. `a == b` with `a is b` fallback -- what I proposed in comment above
5. `a is b or a == b` with fallback to false -- proposed in this PR
6. (leaving out more elaborate schemes like using param id for now)

I don't like 5 because it's sort of logically inconsistent. Consider the `nan` case:

```py
>>> id(float("nan"))
129925236585552
>>> id(float("nan"))
129925236588336
>>> id(float("nan"))
129925236588336
```

So it's not guaranteed to always have the same identity, but sometimes it does. So sometimes it will be a hit and sometimes a miss. With option 1 that's alright, because it's clear identity is always used, so it's basically a user error. But with option 5, when the `is` check is considered merely as an optimization for `==`, now the optimization affects the semantics which is no good IMO. With option 4 this never happens -- `==` and `is` are used for disjoint sets of types so it's logically consistent.

Reply:
Got it, I agree. Will change to 4 then. 👍